### PR TITLE
[perf] Add FA4 CuTe backward support for VSA-256

### PR DESCRIPTION
**Pull request by ita9naiwa**
## Purpose

Enable FA4 backward support for VSA-256 on Blackwell. Related issue/PR: #1354 

## Changes

- Add FA4 CuTe VSA-256 backward by wiring `_flash_attn_bwd` into the existing `_flash_attn_fwd` path.
- Add an autograd bridge supporting both BHSD and BSHD layouts, including variable block sizes.
- Preserve the existing VSA metadata, dispatch, and public interfaces.
- Add backward parity tests, extend the existing VSA benchmark, and update the README.

## Test Plan

```bash

export PYTHONPATH="$PWD/fastvideo-kernel/python${PYTHONPATH:+:$PYTHONPATH}"

python -m pytest -q -s \
  fastvideo-kernel/tests/test_vsa256_forward.py \
  fastvideo-kernel/tests/test_vsa256_forward_vbs.py \
  fastvideo-kernel/tests/test_vsa256_forward_cross.py \
  fastvideo-kernel/tests/test_vsa256_triton.py \
  fastvideo-kernel/tests/test_vsa256_backward.py



# similar to WAN 14B 480p
# 480p: 39,936 tokens, 156 x 256-token blocks, top-k 20
python fastvideo-kernel/benchmarks/bench_vsa.py \
  --block_size 256 --force_triton \
  --batch_size 1 --num_heads 12 --head_dim 128 \
  --q_seq_lens 39936 --topk 20 --warmup 20 --rep 50

python fastvideo-kernel/benchmarks/bench_vsa.py \
  --block_size 256 --use_cute \
  --batch_size 1 --num_heads 12 --head_dim 128 \
  --q_seq_lens 39936 --topk 20 --warmup 20 --rep 50

# similar to WAN 14B 720p
# 720p: 92,160 tokens, 360 x 256-token blocks, top-k 45
python fastvideo-kernel/benchmarks/bench_vsa.py \
  --block_size 256 --force_triton \
  --batch_size 1 --num_heads 12 --head_dim 128 \
  --q_seq_lens 92160 --topk 45 --warmup 20 --rep 50

python fastvideo-kernel/benchmarks/bench_vsa.py \
  --block_size 256 --use_cute \
  --batch_size 1 --num_heads 12 --head_dim 128 \
  --q_seq_lens 92160 --topk 45 --warmup 20 --rep 50
```

## Test Results

```
7 passed in 27.83s

VSA Block-Sparse Attention Benchmark (WRAPPER)
device: NVIDIA B300 SXM6 PC
batch=1, heads=12, head_dim=128, dtype=bf16
block_size=256
NOTE: timings include wrapper overhead (map->index + dispatch).
dispatch: forced Triton (FASTVIDEO_KERNEL_VSA_FORCE_TRITON=1)

================================================================================
q_len=39936, kv_len=39936, num_q_blocks=156, num_kv_blocks=156, topk=20
fwd(wrapper): 2.664 ms  | 471.56 TFLOPs (approx)
bwd(wrapper): 8.442 ms  | 372.04 TFLOPs (approx)

VSA Block-Sparse Attention Benchmark (WRAPPER)
device: NVIDIA B300 SXM6 PC
batch=1, heads=12, head_dim=128, dtype=bf16
block_size=256
NOTE: timings include wrapper overhead (map->index + dispatch).
dispatch: FA4 CuTe

================================================================================
q_len=39936, kv_len=39936, num_q_blocks=156, num_kv_blocks=156, topk=20
fwd(wrapper): 1.407 ms  | 893.19 TFLOPs (approx)
bwd(wrapper): 2.810 ms  | 1117.56 TFLOPs (approx)

VSA Block-Sparse Attention Benchmark (WRAPPER)
device: NVIDIA B300 SXM6 PC
batch=1, heads=12, head_dim=128, dtype=bf16
block_size=256
NOTE: timings include wrapper overhead (map->index + dispatch).
dispatch: forced Triton (FASTVIDEO_KERNEL_VSA_FORCE_TRITON=1)

================================================================================
q_len=92160, kv_len=92160, num_q_blocks=360, num_kv_blocks=360, topk=45
fwd(wrapper): 12.976 ms  | 502.70 TFLOPs (approx)
bwd(wrapper): 41.740 ms  | 390.69 TFLOPs (approx)

VSA Block-Sparse Attention Benchmark (WRAPPER)
device: NVIDIA B300 SXM6 PC
batch=1, heads=12, head_dim=128, dtype=bf16
block_size=256
NOTE: timings include wrapper overhead (map->index + dispatch).
dispatch: FA4 CuTe

================================================================================
q_len=92160, kv_len=92160, num_q_blocks=360, num_kv_blocks=360, topk=45
fwd(wrapper): 5.386 ms  | 1211.16 TFLOPs (approx)
bwd(wrapper): 14.966 ms  | 1089.64 TFLOPs (approx)

```

### Benchmark Summary

| Workload | Backend | Forward | Backward | 
|---|---|---:|---:|
| Wan 14B 480p-like | Triton fallback | 2.664 ms | 8.442 ms | 
| Wan 14B 480p-like | FA4 CuTe | 1.407 ms | 2.810 ms | 
| Wan 14B 720p-like | Triton fallback | 12.976 ms | 41.740 ms | 
| Wan 14B 720p-like | FA4 CuTe | 5.386 ms | 14.966 ms | 

Compared with the Triton fallback, FA4 CuTe provides:

- **480p-like:** **3.00x faster backward**
- **720p-like:** **2.79x faster backward**

## Checklist

- [x] I ran `pre-commit run --all-files` and fixed all issues
- [x] I added or updated tests for my changes
- [x] I updated documentation if needed
- [ ] I considered GPU memory impact of my changes

### CODE REVIEW

**Comment by ita9naiwa** [comment]
@alexzms @SolitaryThinker, since this is a follow-up to #1354, would you mind taking a look when you have a chance?

**Comment by ita9naiwa** [comment]
The CI failure appears to be unrelated to this PR. It comes from the existing `test_backward_asymmetric_q_kv` test, and this PR does not modify the `varlen` VSA path or its tests.

**Comment by alexzms** [comment]
Thank you for this — and welcome. This is a genuinely well-executed piece of work, and it reproduces cleanly: on GB200 (sm_100) I measure 3.06x / 2.76x backward over the Triton fallback at your 480p / 720p shapes, right in line with your B300 numbers.

A few things I want to call out specifically, because they are the parts that are easy to get wrong and you got right. Delegating the autograd bridge to FA4's public `flash_attn_func` rather than hand-wiring `_flash_attn_bwd` is the more robust choice, and it keeps us honest against upstream. Working out that FA4's backward is KV-owned, and transposing the metadata to match while keeping full and partial KV tiles separate so the token-level mask only runs on padded tiles, took real reading of `block_sparsity.py` — it lines up exactly with what `normalize_block_sparse_config_bwd` expects. And the out-of-place change in `video_sparse_attn_bshd` is load-bearing rather than cosmetic: revert just that hunk and the BSHD case fails with `modified by an inplace operation ... output 0 of FlashAttnFuncBackward`. Nice catch. The three commits reverting unrelated formatting churn to keep the diff honest were appreciated too.

I've pushed three commits on top rather than round-tripping them to you:

1. **Docs** — the kernel README pinned `940cd968` (pyproject has been on `82d6441e`) and told you to pre-install `nvidia-cutlass-dsl>=4.5.0`, which leaves you on 4.5.x and then dies at JIT time with `fmax() missing 1 required positional argument`. Unrelated to your change; just what tripped me up reproducing yours.
2. **Merge of main** — conflicts were only yapf churn from #1700 / #6eb95693.
3. **Two fixes + tests** — `video_sparse_attn_h3.py` has the same in-place pattern on FA4's saved output that you fixed in `video_sparse_attn_bshd`, and it breaks the moment this lands. That file arrived after your base, so nothing you did. And the KV-owned backward metadata is now gated on `requires_grad`: inference never reads it, and at Wan-14B 720p shape a no_grad forward goes from 17.23 ms / 1998 MiB to 15.92 ms / 1918 MiB. Tests added for the gated compression branch, partial Q tiles, `q_len != kv_len`, and the H3 backend.

One more thing worth knowing, since you flagged it: the `test_backward_asymmetric_q_kv` failure is real and reproducible, but you were right that it is not yours. `test_vsa_varlen.py` draws all its data from the global RNG and then asserts `max_rel < 0.05`, so any test added ahead of it shifts its inputs — red in a full-suite run, green in isolation. Seeded it in the same commit.

318 passed locally on GB200; the one remaining failure fails identically on an unmodified main. Merging once CI is green — thanks again for taking this on.

**Review by alexzms — APPROVED** [review]
Verified on GB200 (sm_100), the only arch that actually executes this path: 324 passed, CuTe-vs-Triton parity across 128 shape/head/batch/dtype combinations, and backward at 2.6-3.0x over the Triton fallback (61-66% of measured dense bf16 peak). H3 dispatch confirmed end to end with FASTVIDEO_VSA_CUTEDSL=1.

One deployment note before this goes out: the FA4_CUTE_REF bump in docker/Dockerfile means the published fastvideo-dev image keeps the old FA4 until it is rebuilt from main. VSA-128 backward needs the new pin — the old revision only reads block_size[0] on sm_90, so on sm_100 q_subtile_factor comes out as 2 and the 128 metadata is rejected. VSA-256 works on either pin, so H3 is fine on the current image; just don't expect the 128 path to work in a container until the rebuild lands.
